### DISCUSSION TITLE
'NoneType' object is not iterable

### DISCUSSION THREAD

**Discussion by 88arvin** [jsvine/pdfplumber#854] [Q&A]
Earlier, everything was OK and I could extract tables from PDFs, whether they were on one page or spread across numerous pages using the following code.
 
However, after updation some error (TypeError: 'NoneType' object is not iterable) is now appearing only in PDFs that have a table on the first page.

```
with pdfplumber.open(pdf_file) as pdf:
      lst = [p.extract_table() for p in pdf.pages]
flat_list = [item for sublist in lst for item in sublist]
df = pd.DataFrame(flat_list)
```
Does anyone know how to resolve this error?

**Comment by jsvine** [comment]
Hi @88arvin, I believe (without being able to see the actual PDF) that you're likely getting this error because of this line:

```python
flat_list = [item for sublist in lst for item in sublist]
```

That line expects every `sublist` to be a Python list or other iterable; but `sublist` will actually be `None` when a page does not contain a table. There are a few ways you can fix this — here's one:

```python
non_nones = [sublist for sublist in lst if sublist is not None]
flat_list = [item for sublist in non_nones for item in sublist]
```

... which should filter out those `None`s.

**Reply by 88arvin** [reply]
But earlier it was working fine. Anyways I appreciate your response however I did something like this:

```
with pdfplumber.open(pdf_file) as pdf:
        try:
            lst = [p.extract_table() for p in pdf.pages]
            flat_list = [item for sublist in lst for item in sublist]
        except:
            flat_list = pdf.pages[0].extract_table()
```

**Reply by jsvine** [reply]
What version of `pdfplumber` were you using before?
